teps/0115-tekton-catalog-git-based-versioning.md

```diff
@@ -0,0 +1,334 @@
+organization from resource reference. Now that [TEP-0110][tep-0110] is implemented, we can create a new organization and
+versioning contract for Catalogs to address the challenges with the directory-based organization and versioning.
+
+The aim of this TEP is to create a new organization and versioning for Tekton Catalogs, building on the prior work and 
+experiences from [TEP-0003][tep-0003]. In [TEP-0079: Tekton Catalog Support Tiers][tep-0079], we plan to add a Tekton 
+Official Catalog which will use this new organization and versioning contract - therefore, this proposal is a blocker 
+for [TEP-0079][tep-0079].
+
+## Proposal
+
+We propose supporting a flat organization and git-based versioning of Tekton Catalogs.
+
+### Flat Organization
+
+In this organization contract, we'll use a flat structure instead of the directory structure. The example shown above
+with the two `Tasks` (argocd and golang-build) and one `Pipeline` (go-release) would be organized as such:
+
+```shell
+./task/
+  /argocd
+    /README.md
+    /argocd.yaml
+    /samples
+    /tests
+    /OWNERS
+  /golang-build
+    /README.md
+    /golang-build.yaml
+    /samples
+    /tests
+    /OWNERS
+./pipeline/
+  /go-release
+    /README.md
+    /go-release.yaml
+    /samples
+    /tests
+    /OWNERS
+```
+
+Note that there are no changes to the contents of the resources themselves, only the organization of Catalogs.
+
+### Git-Based Versioning
+
+Users can reference specific versions of the resources using git revisions - commits, tags or branches.
+
+#### Git Resolver
+
+[Git Resolver][git-resolver] already has a `"commit"` parameter that can be used to fetch a specific version of a 
+resource, for example [this commit][bp-commit] of buildpacks `Task`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  name: buildpacks
+spec:
+  taskRef:
+    resolver: git
+    resource:
+    - name: url
+      value: https://github.com/tekton/catalog.git
+    - name: path
+      value: task/buildpacks/buildpacks.yaml
+    - name: commit
+      value: d9183e36f00c712b5de59932650c071674ea89b8
+```
+
+[Git Resolver][git-resolver] has a `"branch"` parameter that can be used to get a version of a resource from a
+specific branch, for example [this branch][bp-branch] of buildpacks `Task`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  name: buildpacks
+spec:
+  taskRef:
+    resolver: git
+    resource:
+    - name: url
+      value: https://github.com/tekton/catalog.git
+    - name: path
+      value: task/buildpacks/buildpacks.yaml
+    - name: branch
+      value: main
+```
+
+Note that a user can specify either `"commit"` or `"branch"`, but not both. We propose supporting `"tags"` as well
+as a form of revision. Therefore, we propose removing `"commit"` and `"branch"` from Git Resolver, and replace them
+with `"revision"` which can be a commit, branch or tag.
```

Comment:
prefer the defaulting to "main" if not provided

in the same way, added that "version" in hub resolver will default to the latest version if not provided

Reply:
nit: default to the HEAD instead of main, since this will work regardless of what the default branch of the repo is.

```sh
$ git ls-remote --symref https://github.com/kubernetes/kubernetes HEAD         
ref: refs/heads/master  HEAD
759785ea147bc13945d521eaba4a6592cbc0675f        HEAD
$ git ls-remote --symref https://github.com/tektoncd/pipeline HEAD
ref: refs/heads/main    HEAD
369665cb683e1b09cc4bbf773c39ec78a005431d        HEAD
```